tmc-langs-ant/src/main/java/fi/helsinki/cs/tmc/langs/ant/AntPlugin.java

```diff
@@ -21,23 +32,164 @@ public String getLanguageName() {
+                        new ImmutableMap.Builder<String, byte[]>().put(SpecialLogs.COMPILER_OUTPUT,
+                                java.nio.file.Files.readAllBytes(buildLog.toPath())).build());
+                return false;
+                
+            }
+        } catch (IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private List<String> buildTestScannerArgs(Path path, ClassPath classPath, String... args) {
+        List<String> scannerArgs = new ArrayList<>();
+
+        scannerArgs.add("java");
+        scannerArgs.add("-cp");
+
+        if (classPath == null) {
+            scannerArgs.add(generateClassPath(path).toString());
+        } else {
+            scannerArgs.add(classPath.toString());
+        }
+
+        scannerArgs.add("fi.helsinki.cs.tmc.testscanner.TestScanner");
+        path = path.toAbsolutePath();
+        scannerArgs.add(path.toString() + testDir);
+
+        if (args != null) {
+            for (String arg : args) {
+                scannerArgs.add(arg);
+            }
+        }
+
+        return scannerArgs;
+    }
+
+    private List<String> buildTestRunnerArgs(Path path) {
+        List<String> runnerArgs = new ArrayList<>();
+        List<String> testMethods;
+
+        runnerArgs.add("java");
+        runnerArgs.add("-Dtmc.test_class_dir=" + path.toString() + testDir);
+        runnerArgs.add("-Dtmc.results_file=" + path.toString() + resultsFile);
+        //runnerArgs.add("-Dfi.helsinki.cs.tmc.edutestutils.defaultLocale=" + locale);
+
+        if (endorsedLibsExists(path)) {
+            runnerArgs.add("-Djava.endorsed.dirs=" + createPath(path, "lib", "endorsed"));
+        }
+
+        ClassPath classPath = generateClassPath(path);
+
+        runnerArgs.add("-cp");
+        runnerArgs.add(classPath.toString());
+        runnerArgs.add("fi.helsinki.cs.tmc.testrunner.Main");
+
+        testMethods = startProcess(buildTestScannerArgs(path, classPath, "--test-runner-format"));
+
+        for (String testMethod : testMethods) {
+            runnerArgs.add(testMethod);
+        }
+
+        return runnerArgs;
+    }
+
+    private ClassPath generateClassPath(Path path) {
+        ClassPath classPath = new ClassPath(path.toAbsolutePath());
+        classPath.addDirAndContents(createPath(path, "lib"));
+        classPath.add(createPath(path, "build", "test", "classes"));
+        classPath.add(createPath(path, "build", "classes"));
+
+        return classPath;
+    }
+
+    private boolean endorsedLibsExists(Path path) {
+        File endorsedDir = createPath(path, "lib", "endorsed").toFile();
+        return endorsedDir.exists() && endorsedDir.isDirectory();
+    }
+
+    private Path createPath(Path basePath, String... subDirs) {
+        String path = basePath.toAbsolutePath().toString();
+
+        for (String subDir : subDirs) {
+            path += File.separatorChar + subDir;
+        }
```

Comment:
There shouldn't be any case where you need more than couple of concatenations, the method is just a helper to create paths to the test and lib folders inside the project.